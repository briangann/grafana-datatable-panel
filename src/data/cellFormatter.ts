import {
  dateTime,
  getValueFormat,
} from "@grafana/data";
import _ from 'lodash';
import { DateFormats } from "types";


// Similar to DataLinks, this replaces the value of the panel time ranges for use in url params
export const ReplaceTimeMacros = (timeFrom: string, timeTo: string, content: string) => {
  let newContent = content;
  if (content.match(/\$__from/g)) {
    // replace all occurrences
    newContent = newContent.replace('$__from', timeFrom);
  }
  if (content.match(/\$__to/g)) {
    // replace all occurrences
    newContent = newContent.replace('$__to', timeTo);
  }
  if (content.match(/\$__keepTime/g)) {
    // replace all occurrences
    const keepTime = `from=${timeFrom}&to=${timeTo}`;
    newContent = newContent.replace('$__keepTime', keepTime);
  }
  return newContent;
};

export const TimeFormatter = (timeZone: string, timestamp: number, timestampFormat: string): string => {
  const timestampFormatted =
    timeZone === 'utc'
      ? dateTime(timestamp)
        .utc()
        .format(timestampFormat)
      : dateTime(timestamp).format(timestampFormat);
  return timestampFormatted;
}


/**
 * [formatColumnValue description]
 * @param  {[type]} colIndex [description]
 * @param  {[type]} rowIndex [description]
 * @param  {[type]} value    [description]
 * @return {[type]}          [description]
 */
export const FormatColumnValue = (data: any, colIndex: any, rowIndex: any, value: any, valueType: string, timeFrom: string, timeTo: string) => {
  //console.log(valueType);
  // the valueFormatter for "time" expects a string already formatted
  // this needs to check if the value is an epoch, and convert that...

  // if is an epoch and type time (numeric string and len > 12)
  if ((valueType === 'time') && !isNaN(value as any)) {
    const parsed = parseInt(value, 10);
    // TODO: fix this
    // May need to adjust for local time zone here
    const fo = TimeFormatter('utc', parsed, DateFormats[5].value);
    return fo;
  }

  if (valueType === 'other') {
    return JSON.stringify(value);
  }
  const aFormatter = getValueFormat(valueType);
  let formatted = aFormatter(value).text;
  // TODO: fix this!
  if (/\$__cell_\d+/.exec(value)) {
    // eslint-disable-next-line no-debugger
    //debugger;
    for (let i = data.columns - 1; i >= 0; i--) {
      formatted = formatted.replace(`$__cell_${i}`, data.rows[rowIndex][i]);
    }
  }
  return formatted;
};
