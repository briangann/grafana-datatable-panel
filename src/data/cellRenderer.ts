import {
  dateTime,
  Field,
  formattedValueToString,
  getValueFormat,
  GrafanaTheme2,
  TimeRange,
} from "@grafana/data";

import _ from 'lodash';
import { DateFormats } from "types";
import { DTColumnType } from "./types";
import moment from 'moment-timezone';
import { ColumnStyleItemType } from "components/options/columnstyles/types";

// Similar to DataLinks, this replaces the value of the panel time ranges for use in url params
export const ReplaceTimeMacros = (timeFrom: string, timeTo: string, content: string) => {
  let newContent = content;
  if (content.match(/\$__from/g)) {
    // replace all occurrences
    newContent = newContent.replace('$__from', timeFrom);
  }
  if (content.match(/\$__to/g)) {
    // replace all occurrences
    newContent = newContent.replace('$__to', timeTo);
  }
  if (content.match(/\$__keepTime/g)) {
    // replace all occurrences
    const keepTime = `from=${timeFrom}&to=${timeTo}`;
    newContent = newContent.replace('$__keepTime', keepTime);
  }
  return newContent;
};

/**
 * [TimeFormatter description]
 *
 * @param   {string}  timeZone         [timeZone description]
 * @param   {number}  timestamp        [timestamp description]
 * @param   {string}  timestampFormat  [timestampFormat description]
 *
 * @return  {string}                   [return description]
 */
export const TimeFormatter = (timeZone: string, timestamp: number, timestampFormat: string): string => {
  if (timeZone === 'utc') {
    const timestampFormatted = dateTime(timestamp)
    .utc()
    .format(timestampFormat);
    return timestampFormatted;
  }
  // this appears to be bugged
  //
  // const timestampFormatted = dateTimeForTimeZone(timeZone, timestamp);
  // console.log(timestampFormatted.toISOString(true));

  // TODO: bug
  // when timezone is browser, moment.tz can't be used
  //
  let formattedWithTimezone = dateTime(timestamp).format(timestampFormat);
  if (timeZone !== 'browser') {
    formattedWithTimezone = moment.tz(
    dateTime(timestamp).utc().toISOString(true),
    timestampFormat, timeZone).format(timestampFormat);
  }
  return formattedWithTimezone;
}

/**
 * [FormatColumnValue description]
 *
 * @param   {Field}          field      [field description]
 * @param   {number}         colIndex   [colIndex description]
 * @param   {number}         rowIndex   [rowIndex description]
 * @param   {any}            value      [value description]
 * @param   {string}         valueType  [valueType description]
 * @param   {string}         timeFrom   [timeFrom description]
 * @param   {string}         timeTo     [timeTo description]
 * @param   {GrafanaTheme2}  theme      [theme description]
 *
 * @return  {string}                    [Formatted Value]
 */
export const FormatColumnValue = (userTimeZone: string, columnStyle: ColumnStyleItemType|null, field: Field, colIndex: number, rowIndex: number, value: any, valueType: string, timeFrom: string, timeTo: string, theme: GrafanaTheme2): string => {
  // if is an epoch and type time (numeric string and len > 12)
  if ((valueType === 'time') && !isNaN(value as any)) {
    const parsed = parseInt(value, 10);
    let dateFormat = DateFormats[5].value;
    if (columnStyle && columnStyle.dateFormat) {
      dateFormat = columnStyle.dateFormat;
    }
    // timezone comes from user preferences
    const formatted = TimeFormatter(userTimeZone, parsed, dateFormat);
    return formatted;
  }

  // value type String

  if (valueType === 'other') {
    return JSON.stringify(value);
  }
  const aFormatter = getValueFormat(field.config.unit);

  let maxDecimals = 4;
  if (field.config.decimals !== undefined && field.config.decimals !== null) {
      maxDecimals = field.config.decimals;
  }

  let formatted = formattedValueToString(aFormatter(value, maxDecimals));
  return formatted;
};

// TODO: this is not complete
export const ProcessMacrosForClickthrough = (
  columnStyle: ColumnStyleItemType|null,
  columns: any,
  rows: any,
  colIndex: number,
  rowIndex: number,
  value: any,
  valueType: string,
  timeRange: TimeRange) => {
  // check for $__keepTime
  // eslint-disable-next-line no-debugger
  debugger;
  if (!columnStyle?.clickThrough) {
    return null;
  }
  const keepTimeRegex = RegExp(/\$__keepTime/);
  let clickThrough = columnStyle?.clickThrough;
  if (clickThrough.match(keepTimeRegex)) {
    clickThrough = clickThrough.replace(`$__keepTime`, `from=${timeRange.raw.from}&to=${timeRange.raw.to}`);
  }
  const aRegex = RegExp(/\$__cell_\d+/);
  if (clickThrough.match(aRegex)) {
    for (let i = columns.length - 1; i >= 0; i--) {
      // this is text content, whatever is in the row should be formatted
      // eslint-disable-next-line no-debugger
      debugger;
      const cellContent = rows[rowIndex].mean;
      console.log(cellContent);
      clickThrough = clickThrough.replace(`$__cell_${i}`, rows[rowIndex].mean);
    }
  }
  return clickThrough;
}

export const ApplyUnitsAndDecimals = (columns: DTColumnType[], rows: any[]) => {
  for (const item of columns) {
    const aStyle = item.columnStyle;
    if (aStyle) {
      let decimals = Number(item.columnStyle?.decimals).valueOf();
      let unit = aStyle?.unitFormat || 'short';
      let colName = item.data;
      for (const aRow of rows) {
        let value = aRow[colName];
        const formatted = applyFormat(value, decimals, unit)
        aRow[colName] = formatted;
      }
    }
  }
  return {columns, rows};
}

export const applyFormat = (value: any, maxDecimals: number, unitFormat: string) => {
  let valueFormatted = '';
  let valueRounded = '';
  let valueRoundedAndFormatted = '';
  const formatFunc = getValueFormat(unitFormat);
  if (formatFunc) {
    const decimals: number = maxDecimals;
    const formatted = formatFunc(value, decimals);

    valueFormatted = formatted.text;
    valueRoundedAndFormatted = roundValue(value, decimals) || value;
    valueRounded= roundValue(value, decimals) || value;
    // spaces are included with the formatFunc
    if (formatted.suffix) {
      valueFormatted += formatted.suffix;
      valueRoundedAndFormatted += formatted.suffix;
    }
    if (formatted.prefix) {
      valueFormatted = formatted.prefix + valueFormatted;
      valueRoundedAndFormatted = formatted.prefix + valueRoundedAndFormatted;
    }
  }
  return (
    {
      valueRaw: value,
      valueFormatted: valueFormatted,
      valueRounded: valueRounded,
      valueRoundedAndFormatted: valueRoundedAndFormatted,
    }
  )
}

const roundValue = (num: number, decimals: number): number | null => {
  if (num === null) {
    return null;
  }
  const n = Math.pow(10, decimals);
  const formatted = (n * num).toFixed(decimals);
  return Math.round(parseFloat(formatted)) / n;
};
