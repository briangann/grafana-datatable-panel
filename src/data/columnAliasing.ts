import { ColumnAliasField } from "types";
import { DTColumnType } from "./types";

const getColumnAlias = (columnName: any, columnAliases: ColumnAliasField[]) => {
  // default to the columnName
  let columnAlias = columnName;
  if (columnAliases.length > 0) {
    for (let i = 0; i < columnAliases.length; i++) {
      if (columnAliases[i].name === columnName) {
        columnAlias = columnAliases[i].alias;
        break;
      }
    }
  }
  return columnAlias;
}

export const ApplyColumnAliases = (dtColumns: DTColumnType[], columnAliases: ColumnAliasField[]): DTColumnType[] => {
  for (let i = 0; i < dtColumns.length; i++) {
    const anAlias = getColumnAlias(dtColumns[i].title, columnAliases);
    if (anAlias) {
      dtColumns[i].title = anAlias;
    }
  }
  return dtColumns;
}
