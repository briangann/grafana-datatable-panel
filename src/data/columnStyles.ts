import { ColumnStyleItemType } from "components/options/columnstyles/types";
import { DTColumnType } from "./types";

export const ApplyColumnStyles = (columns: DTColumnType[], columnStyles: ColumnStyleItemType[]) => {
  for (const item of columns) {
    for (let index = 0; index < columnStyles.length; index++) {
      const aStyle = columnStyles[index];
      // convert to regexp

      let expression = `${aStyle.nameOrRegex}`;
      // remove leading and ending slashes
      if (expression.startsWith(`/`)) {
        expression = expression.slice(1);
      }
      if (expression.endsWith(`/`)) {
        expression = expression.slice(0, -1);
      }
      const rx = new RegExp(expression);
      //const matches = rx.test(item.title);
      //console.log(matches);
      if (item.title.match(rx)) {
        //console.log(`ApplyColumnStyles: style match found for column ${item}`);
        // set the column style for the item, to be used in rendering
        item.columnStyle = aStyle;
        // matched move on to next column
        break;
      }
    }
  };
  return columns;
};
