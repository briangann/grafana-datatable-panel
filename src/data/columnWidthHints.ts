import { ColumnWidthHint } from "types";
import { DTColumnType } from "./types";

const getColumnWidthHint = (name: string, columnWidthHints: ColumnWidthHint[]) => {
  for (let i = 0; i < columnWidthHints.length; i++) {
    if (columnWidthHints[i].name === name) {
      return columnWidthHints[i].width;
    }
  }
  return null;
}

export const ApplyColumnWidthHints = (dtColumns: DTColumnType[], columnWidthHints: ColumnWidthHint[]): DTColumnType[] => {
  for (let i = 0; i < dtColumns.length; i++) {
    const aHint = getColumnWidthHint(dtColumns[i].title!, columnWidthHints);
    if (aHint) {
      dtColumns[i].widthHint = aHint;
    }
  }
  return dtColumns;
}
