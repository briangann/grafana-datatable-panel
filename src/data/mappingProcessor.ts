import { ValueMapping } from "@grafana/data";
import { FormattedColumnValue } from "./types";
import { getValueMappingResult } from "./valueMappings";

export const GetMappings = (fieldConfigMappings: ValueMapping[] | undefined, dataMappings: ValueMapping[] | undefined) => {
  return fieldConfigMappings && fieldConfigMappings.length > 0 ? fieldConfigMappings : dataMappings;
}

export const ApplyMappings = (value: FormattedColumnValue, mappings: any) => {
  const aValue = getValueMappingResult(mappings, value.valueRaw);
  if (aValue !== null) {
    //console.log(`ApplyMappings matched, mapped value = ` + JSON.stringify(aValue));
    value.valueFormatted = aValue.text || '';
    return value;
  }
  return null;
}
