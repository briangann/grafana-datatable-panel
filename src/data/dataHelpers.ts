// FieldType across runtimes are not working
import {
  DataFrame,
  Field,
  FieldType,
  GrafanaTheme2
} from '@grafana/data';
import { ApplyUnitsAndDecimals, FormatColumnValue } from 'data/cellRenderer';
import { ApplyGrafanaOverrides } from './overrides';
import { ConfigColumnDefs } from 'datatables.net';
import { ColumnStyleColoring, ColumnStyleType } from 'types';
import { DTColumnType } from './types';
import { ColumnStyleItemType } from 'components/options/columnstyles/types';
import { ApplyColumnStyles } from './columnStyles';
import { DTData } from 'components/DataTablePanel';

function normalizeFieldName(field: string) {
  return field
    .replace(/ /g, '_')
    .replace(/[^a-zA-Z0-9_]/g, '')
    .toLowerCase();
}

export const DataFrameToDisplay = (frames: DataFrame[]) => {
  const frame = frames[0];
  const valueFields: Field[] = [];
  let newestTimestamp = 0;
  for (const aField of frame.fields) {
    if (aField.type === FieldType.number) {
      valueFields.push(aField);
    }
    else if (aField.type === FieldType.time) {
      // get the "newest" timestamp from data
      // check if timestamp is 0
      let timestampIndex = aField.values.length - 1;
      let aTimestamp = aField.values[timestampIndex];
      if (newestTimestamp === 0) {
        newestTimestamp = aTimestamp;
      }
      // check if data timestamp is newer
      if (aTimestamp > newestTimestamp) {
        newestTimestamp = aTimestamp;
      }
    }
  }
  if (newestTimestamp === 0) {
    // use current time if none is found
    newestTimestamp = new Date().getTime()
  }
};


export const ConvertDataFrameToDataTableFormat = (
  dataFrames: DataFrame[],
  userTimeZone: string,
  alignNumbersToRightEnabled: boolean,
  rowNumbersEnabled: boolean,
  columnStyles: ColumnStyleItemType[],
  theme: GrafanaTheme2): { columns: DTColumnType[]; rows: any[] } => {
  DataFrameToDisplay(dataFrames);
  dataFrames = ApplyGrafanaOverrides(dataFrames, theme);
  const dataFrame = dataFrames[0];
  let columns: DTColumnType[] = dataFrame.fields.map((field) => {
    const columnClassName = getColumnClassName(alignNumbersToRightEnabled, field.type as string)
    return {
      title: field.name,
      data: normalizeFieldName(field.name),
      type: field.type as string,
      className: columnClassName,
      fieldConfig: field.config,
      columnStyle: null,
      widthHint: '',
    };
  });
  ApplyColumnStyles(columns, columnStyles);

  const rows = [] as any[];

  for (let i = 0; i < dataFrame.length; i++) {
    const row = {};
    for (let j = 0; j < columns.length; j++) {
      const aColumn = columns[j];
      const frameFields = dataFrame.fields[j];
      let value = frameFields.values[i];
      const valueType = frameFields.type;
      if (valueType !== 'string') {
        value = FormatColumnValue(userTimeZone, aColumn.columnStyle, frameFields, j, i, value, valueType, "timeFrom", "timeTo", theme);
      }
      const colName = columns[j].data;
      // @ts-ignore
      row[colName] = value;
    }
    rows.push(row as any);
  }
  if (rowNumbersEnabled) {
    columns.unshift({
      title: 'Row',
      data: 'rowNumber',
      type: 'number',
      className: '',
      fieldConfig: {},
      columnStyle: null,
      widthHint: '',
    });
    for (let i = 0; i < dataFrame.length; i++) {
      // @ts-ignore
      rows[i].rowNumber = i;
    }
  }
  ApplyUnitsAndDecimals(columns, rows);
  return { columns, rows };
}

// const doStuff = (dtColumns: DTColumnType[], cell: any, fontSizePercent: any, colIndex: number) => {
//   // eslint-disable-next-line no-debugger
//   debugger;
//   $(cell).css('font-size', fontSizePercent);
//   //const aColumn = dtColumns[colIndex];
//   let colorMode = dtColumns[colIndex].columnStyle?.colorMode;
//   console.log(colorMode);
// }

export const BuildColumnDefs = (
  emptyDataEnabled: boolean,
  emptyDataText: string,
  rowNumbersEnabled: boolean,
  fontSizePercent: string,
  alignNumbersToRightEnabled: boolean,
  dtData: DTData): ConfigColumnDefs[] => {

  const columnDefs: ConfigColumnDefs[] = [];
  let rowNumberOffset = 0;
  for (let i = 0; i < dtData.Columns.length; i++) {
    let columnType = dtData.Columns[i].type!;
    let columnClassName = getColumnClassName(alignNumbersToRightEnabled, columnType)
    // column type "date" is very limited, and overrides our formatting
    // best to use our format, then the "raw" epoch time as the sort ordering field
    // https://datatables.net/reference/option/columns.type
    if (columnType === 'time') {
      columnType = 'number';
    }
    if (columnType === 'number' && alignNumbersToRightEnabled) {
      columnClassName = 'dt-right'; // any reason not to align numbers right?
    }
    // if we did not get a type prop from grafana at all,
    // check at least if it's a number to have DT sort properly
    if (columnType !== undefined && dtData.Rows[0] && (typeof dtData.Rows[0][i]) === 'number') {
      columnType = 'number';
    }

    dtData.Columns[i].className = columnClassName;
    // NOTE: the width below is a "hint" and will be overridden as needed,
    // this lets most tables show timestamps with full width
    const columnDefDict: any = {
      width: dtData.Columns[i].widthHint,
      targets: i + rowNumberOffset,
      defaultContent: emptyDataEnabled ? emptyDataText : '',
      data: function (row: any, type: any, set: any, meta: any) {
        if (type === undefined) {
          return null;
        }
        const idx = meta.col;
        if (row[idx]?.display !== undefined) {
          return row[idx].display;
        }
        return null;
      },
      render: function (data: any, type: any, val: any, meta: any) {
        if (type === undefined) {
          return null;
        }
        const aColumn = dtData.Columns[meta.col];
        if (aColumn === undefined) {
          console.log('aColumn undefined');
          return null;
        }
        // //console.log(aColumn);
        // let colorMode = aColumn.columnStyle?.colorMode;
        // //console.log(colorMode);
        // if (colorMode === undefined) {
        //   console.log('colorMode undefined');
        //   //return null;
        // }
        // // eslint-disable-next-line no-debugger
        // debugger;
        // TODO: call render function vs just returning formatted value
        const idx = meta.col;
        if (type === 'type') {
          return val[idx];
        }
        let returnValue = val[meta.col];
        if (returnValue.valueFormatted) {
          return returnValue.valueFormatted;
        }
        return returnValue;
      },
      createdCell: function(cell: any, cellDataAlwaysEmpty: any, rowData: any, rowIndex: number, colIndex: number) {
        // cellData is empty since we use render()
        // eslint-disable-next-line no-debugger
        debugger;
        const aColumn = dtData.Columns[colIndex];
        const colorMode = aColumn.columnStyle?.colorMode;
        console.log(aColumn);
        // set the fontsize for the cell
        $(cell).css('font-size', fontSizePercent);
        // orthogonal sort requires getting cell data differently
        const cellContent = $(cell).html();
        // hidden columns have null data
        if (cellContent === null) {
          return;
        }
        // undefined types should have numerical data, any others are already formatted
        let actualColumn = colIndex;
        if (rowNumbersEnabled) {
          actualColumn -= 1;
        }
        // for coloring rows, get the "worst" threshold
        let rowColor = null;
        let color = null;
        let rowColorIndex = null;
        let rowColorData = null;
        //colorMode = aColumn.columnStyle?.colorMode;
        if (colorMode === ColumnStyleColoring.Row) {
          // run all of the rowData through threshold check, get the "highest" index
          // and use that for the entire row
          if (rowData === null) {
            return;
          }
          rowColorIndex = -1;
          rowColorData = null;
          rowColor = 'teal'; // will stand out with a default like this, useful for debugging
          // this should be configurable...
          color = 'white';
          for (let columnNumber = 0; columnNumber < dtData.Columns.length; columnNumber++) {
            let aColumnStyle = dtData.Columns[columnNumber].columnStyle;
            // need the style to get the color
            if (!aColumnStyle) {
              //console.log(`no style found for column ${columnNumber}`);
              continue;
            }
            // only process color values for numbers
            if (aColumnStyle.styleItemType !== ColumnStyleType.Metric) {
              continue;
            }
            rowColorData = getCellColors(
              aColumnStyle,
              columnNumber,
              rowData[columnNumber + rowNumberOffset]
            );
            if (!rowColorData) {
              continue;
            }

            if (rowColorData.bgColorIndex !== null) {
              if (rowColorData.bgColorIndex > rowColorIndex) {
                rowColorIndex = rowColorData.bgColorIndex;
                rowColor = rowColorData.bgColor;
              }
            }
          }
          // style the entire row (the parent of the td is the tr)
          const fmtColors = 'color: ' + color + ' !important;' +
                'background-color: ' + rowColor + ' !important;';
          $(cell.parentNode)
            .children()
            .attr('style', function(i,s) { return s + fmtColors });
        }

        if (colorMode === ColumnStyleColoring.RowColumn) {
          // run all of the rowData through threshold check, get the "highest" index
          // and use that for the entire row
          if (rowData === null) {
            return;
          }
          rowColorIndex = -1;
          rowColorData = null;
          rowColor = 'blue'; // colorState.rowcolumn;
          // this should be configurable...
          color = 'white';
          for (let columnNumber = 0; columnNumber < dtData.Columns.length; columnNumber++) {
            if (dtData.Columns[columnNumber].type === undefined) {
              if (dtData.Columns[columnNumber].columnStyle !== null) {
                let aColumnStyle = dtData.Columns[columnNumber].columnStyle;
                // need the style to get the color
                if (!aColumnStyle) {
                  //console.log(`no style found for column ${columnNumber}`);
                  continue;
                }
                // only process color values for numbers
                if (aColumnStyle.styleItemType !== ColumnStyleType.Metric) {
                  continue;
                }
                rowColorData = getCellColors(
                  dtData.Columns[columnNumber].columnStyle!,
                  columnNumber,
                  rowData[columnNumber + rowNumberOffset]
                );
              }
              if (!rowColorData) {
                continue;
              }
              if (rowColorData.bgColorIndex !== null) {
                if (rowColorData.bgColorIndex > rowColorIndex) {
                  rowColorIndex = rowColorData.bgColorIndex;
                  rowColor = rowColorData.bgColor;
                }
              }
            }
          }
          // style the rowNumber and Timestamp column
          // the cell colors will be determined in the next phase
          if (dtData.Columns[0].type !== undefined) {
            const children = $(cell.parentNode).children();
            let aChild = children[0];
            $(aChild).css('color', color);
            if (rowColor) {
              $(aChild).css('background-color', rowColor);
            }
            // the 0 column contains the row number, if they are enabled
            // then the above just filled in the color for the row number,
            // now take care of the timestamp
            if (rowNumbersEnabled) {
              aChild = children[1];
              $(aChild).css('color', color);
              if (rowColor) {
                $(aChild).css('background-color', rowColor);
              }
            }
          }
        }

        // Process cell coloring
        // Two scenarios:
        //    1) Cell coloring is enabled, the above row color is skipped
        //    2) RowColumn is enabled, the above row color is process, but we also
        //    set the cell colors individually
        // if (aColumn.columnStyle && aColumn.columnStyle.styleItemType !== ColumnStyleType.Number) {
        //   return;
        // }
        const colorData = getCellColors(aColumn.columnStyle, actualColumn, cellContent);
        if (!colorData) {
          return;
        }
        if (colorMode === ColumnStyleColoring.Cell || colorMode === ColumnStyleColoring.RowColumn) {
          if (colorData && colorData.color !== null) {
            $(cell).css('color', colorData.color);
          }
          if (colorData && colorData.bgColor !== null) {
            $(cell).css('background-color', colorData.bgColor);
          }
        } else if (colorData.color) { // TODO: fix this
          if (colorData && colorData.color !== null) {
            $(cell).css('color', colorData.color);
          }
        }
      },
    };
    //let ignoreNullValues = this.getColumnIgnoreNullValue(i);
    //if (ignoreNullValues) {
    //  columnDefDict.defaultContent = '-';
    //}
    columnDefs.push(columnDefDict);
  }
  return columnDefs;
};

const getColumnClassName = (alignNumbersToRightEnabled: boolean, columnType: string) => {
  let columnClassName = '';

  // column type "date" is very limited, and overrides our formatting
  // best to use our format, then the "raw" epoch time as the sort ordering field
  // https://datatables.net/reference/option/columns.type
  if (columnType === 'time') {
    columnType = 'number';
  }
  if (columnType === 'number' && alignNumbersToRightEnabled) {
    columnClassName = 'dt-right'; // any reason not to align numbers right?
  }
  if (columnType === 'string') {
    columnClassName = 'dt-right';
  }
  return columnClassName;
}

const getCellColors = (aColumnStyle: ColumnStyleItemType | null, columnNumber: any, cellData: any) => {
  if (aColumnStyle === null || cellData === null || cellData === undefined) {
    return null;
  }
  let useData = cellData;
  if (cellData.valueRaw) {
    useData = cellData.valueRaw;
    //console.log(`using valueRaw... ${cellData.valueRaw}`);
  } else {
    // this should not happen
    return null;
  }
  const items = useData.split(/([^0-9.,]+)/);
  // only color cell if the content is a number?
  let bgColor = null;
  let bgColorIndex = null;
  let color = null;
  let colorIndex = null;
  let value = null;
  // check if the content has a numeric value after the split
  if (!isNaN(Number(items[0]))) {
    value = parseFloat(items[0].replace(',', '.'));
  }

  if (aColumnStyle && aColumnStyle.colorMode != null && aColumnStyle.thresholds.length > 0) {
    // check color for either cell or row
    if (aColumnStyle.colorMode === ColumnStyleColoring.Cell ||
      aColumnStyle.colorMode === ColumnStyleColoring.Row ||
      aColumnStyle.colorMode === ColumnStyleColoring.RowColumn) {
      if (value !== null && !isNaN(value)) {
        bgColor = GetColorForValue(value, aColumnStyle);
        bgColorIndex = GetColorIndexForValue(value, aColumnStyle);
      }
      color = 'white';
    }
    // just the value color is set
    if (aColumnStyle.colorMode === ColumnStyleColoring.Value) {
      if (value !== null && !isNaN(value)) {
        color = GetColorForValue(value, aColumnStyle);
        colorIndex = GetColorIndexForValue(value, aColumnStyle);
      }
    }
  }
  return {
    bgColor: bgColor,
    bgColorIndex: bgColorIndex,
    color: color,
    colorIndex: colorIndex,
  };
};

export const GetColorForValue = (value: number, style: ColumnStyleItemType) => {
  if (!style.thresholds) {
    return null;
  }
  let color = style.thresholds[0].color;
  //console.log(`GetColorForValue: checking value ${value}`);
  for (let i = style.thresholds.length - 1; i > 0; i--) {
    const checkValue = style.thresholds[i].value;
    //console.log(`GetColorForValue: checking value ${value} >= ${checkValue}`);
    if (value >= checkValue) {
      //console.log(`GetColorForValue: value ${value} IS >= ${checkValue} return color index ${i}`);
      color = style.thresholds[i].color;
      // found highest match
      break;
    }
  }
  return color;
};

// to determine the overall row color, the index of the threshold is needed
export const GetColorIndexForValue = (value: any, style: any) => {
  if (!style.thresholds) {
    return null;
  }
  let colorIndex = 0;
  for (let i = style.thresholds.length - 1; i > 0; i--) {
    if (value >= style.thresholds[i].value) {
      colorIndex = i;
      break;
    }
  }
  return colorIndex;
};
