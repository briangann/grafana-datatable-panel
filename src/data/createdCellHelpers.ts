import { DTData } from "components/DataTablePanel";
import { getCellColors } from "./dataHelpers";
import { DTColumnType, FormattedColumnValue } from "./types";
import { ProcessClickthrough } from "./cellRenderer";
import { ColumnStyleItemType, ColumnStyles } from "components/options/columnstyles/types";
import { TimeRange } from "@grafana/data";


export const processRowStyle = (
  cell: any,
  rowData: any,
  dtData: DTData,
  rowNumberOffset: number) => {

  let rowColorIndex = -1;
  let rowColorData = null;
  let rowColor = null;
  // this should be configurable...
  let color = 'white';

  for (let columnNumber = 0; columnNumber < dtData.Columns.length; columnNumber++) {
    let aColumnStyle = dtData.Columns[columnNumber].columnStyle;
    // need the style to get the color
    if (!aColumnStyle) {
      // no style set, skip
      continue;
    }
    // only process color values for numbers
    if (aColumnStyle.activeStyle !== ColumnStyles.METRIC) {
      continue;
    }
    rowColorData = getCellColors(
      aColumnStyle,
      columnNumber,
      rowData[columnNumber + rowNumberOffset]
    );
    if (!rowColorData) {
      continue;
    }

    if (rowColorData.bgColorIndex !== null) {
      if (rowColorData.bgColorIndex > rowColorIndex) {
        rowColorIndex = rowColorData.bgColorIndex;
        rowColor = rowColorData.bgColor;
      }
    }
  }
  // style the entire row (the parent of the td is the tr)
  const fmtColors = 'color: ' + color + ' !important;' +
    'background-color: ' + rowColor + ' !important;';
  $(cell.parentNode)
    .children()
    .attr('style', function (i, s) { return s + fmtColors });

}

export const processRowColumnStyle = (
  cell: any,
  rowData: any,
  columnsInCellData: DTColumnType[],
  rowNumbersEnabled: boolean,
  rowNumberOffset: number) => {

  let rowColorIndex = -1;
  let rowColorData = null;
  let rowColor = null;
  // this should be configurable...
  let color = 'white';
  for (let columnNumber = 0; columnNumber < columnsInCellData.length; columnNumber++) {
    if (columnsInCellData[columnNumber].type !== undefined) {
      if (columnsInCellData[columnNumber].columnStyle !== null) {
        let aColumnStyle = columnsInCellData[columnNumber].columnStyle;
        // need the style to get the color
        if (!aColumnStyle) {
          continue;
        }
        // only process color values for numbers
        if (aColumnStyle.activeStyle !== ColumnStyles.METRIC) {
          continue;
        }
        rowColorData = getCellColors(
          columnsInCellData[columnNumber].columnStyle!,
          columnNumber,
          rowData[columnNumber + rowNumberOffset]
        );
      }
      if (!rowColorData) {
        continue;
      }
      if (rowColorData.bgColorIndex !== null) {
        if (rowColorData.bgColorIndex > rowColorIndex) {
          rowColorIndex = rowColorData.bgColorIndex;
          rowColor = rowColorData.bgColor;
        }
      }
    }
  }

  for (let columnNumber = 0; columnNumber < columnsInCellData.length; columnNumber++) {
    if (!columnsInCellData[columnNumber].columnStyle) {
      const children = $(cell.parentNode).children();
      let aChild = children[columnNumber];
      $(aChild).css('color', color);
      if (rowColor) {
        // ugly but it works..
        $(aChild)[0].style.setProperty('background-color', rowColor, 'important');
      }
    }
  }
}

export const ProcessStringValueStyle = (
  columnStyle: ColumnStyleItemType | null,
  columnsInCellData: DTColumnType[],
  rowData: any,
  valueFormatted: FormattedColumnValue,
  timeRange: TimeRange): string | null => {

  const processedURL = ProcessClickthrough(
    columnStyle,
    columnsInCellData,
    rowData,
    valueFormatted,
    timeRange);
  if (processedURL !== undefined) {
    return processedURL;
  }
  return null;
};
