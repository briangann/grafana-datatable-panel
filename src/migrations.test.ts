import { PanelModel } from '@grafana/data';

import {
  DatatablePanelMigrationHandler,
} from './migrations';

describe('Datatable -> DatatableV2 migrations', () => {
  it('only migrates old datatable', () => {
    const panel = {} as PanelModel;
    const options = DatatablePanelMigrationHandler(panel);
    expect(options).toEqual({});
  });

  it('migrates old datatable config', () => {
    const panel = {} as PanelModel;
    panel.options = {};
    const options = DatatablePanelMigrationHandler(panel);
    expect(options).toMatchSnapshot();
  });

});
