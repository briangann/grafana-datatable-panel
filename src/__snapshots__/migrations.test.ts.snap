// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datatable -> DatatableV2 migrations migrates old datatable config 1`] = `{}`;
